chromium_src/chrome/browser/ssl/https_only_mode_navigation_throttle.cc

```diff
@@ -0,0 +1,136 @@
+#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ssl/https_only_mode_controller_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/stateful_ssl_host_state_delegate.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace {
+
+// Tor is slow and needs a longer fallback delay
+constexpr base::TimeDelta kTorFallbackDelay = base::Seconds(20);
+
+bool ShouldUpgradeToHttps(content::NavigationHandle* handle) {
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return false;
+  }
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  const GURL& url = handle->GetURL();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  return brave_shields::ShouldUpgradeToHttps(
+      map, url, g_brave_browser_process->https_upgrade_exceptions_service());
+}
+
+bool IsTor(content::NavigationHandle* handle) {
+  auto* context = handle->GetWebContents()->GetBrowserContext();
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsTor();
+}
+
+}  // namespace
+
+#define WillFailRequest WillFailRequest_ChromiumImpl
+#define GetBoolean(PREF_NAME) \
+  GetBooleanOr(PREF_NAME, ShouldUpgradeToHttps(handle))
+#define SetNavigationTimeout(DEFAULT_TIMEOUT)                         \
+  SetNavigationTimeout(IsTor(navigation_handle()) ? kTorFallbackDelay \
+                                                  : DEFAULT_TIMEOUT)
+
+#include "src/chrome/browser/ssl/https_only_mode_navigation_throttle.cc"
+
+#undef WillFailRequest
+#undef GetBoolean
+#undef SetNavigationTimeout
+
+// Called if there is a non-OK net::Error in the completion status.
+content::NavigationThrottle::ThrottleCheckResult
+HttpsOnlyModeNavigationThrottle::WillFailRequest() {
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return WillFailRequest_ChromiumImpl();
+  }
+
+  auto* handle = navigation_handle();
+  auto* contents = handle->GetWebContents();
+  const GURL& request_url = handle->GetURL();
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  Profile* profile = Profile::FromBrowserContext(context);
+  auto* prefs = profile->GetPrefs();
+  if (profile->IsTor() ||
+      (prefs && prefs->GetBoolean(prefs::kHttpsOnlyModeEnabled)) ||
+      brave_shields::ShouldForceHttps(map, request_url)) {
+    return WillFailRequest_ChromiumImpl();
+  }
+
+  const net::SSLInfo info = handle->GetSSLInfo().value_or(net::SSLInfo());
+  int cert_status = info.cert_status;
+  if (!net::IsCertStatusError(cert_status) &&
```

Comment:
I checked a couple of HSTS sites and didn't see any fallback. For example, I visited
http://arthuredelstein.net, which uses HSTS and has upgrades enabled. What I see is that there's a 307 internal redirect to https://arthuredelstein.net because of HSTS.

I also confirmed that if I enter the address https://insecure.arthuredelstein.net (note the secure scheme): that I see a "Your connection is not private" error, but I don't see a fallback.

But to really test your example, I think I need a website that presents Strict-Transport-Security header on the first visit, and then sends a TLS error on the second visit. That seems possible but will take time. :)

But in the meantime I confirmed that [this code](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ssl/https_only_mode_upgrade_interceptor.cc;l=133;drc=9269d4ed0dde3bcd3837f75a50a2944674555186) in Chromium prevents the `HttpsOnlyModeUpgradeInterceptor` from upgrading the navigation. That means Chromium relies instead on the HSTS-based redirect to upgrade. And the fallback code in my PR [explicitly prevents](https://github.com/brave/brave-core/pull/16521/files#diff-f1fd0bc31e2ee294652d5a178d39a51b0b46b51fb1ebc452f21e8342eaf7da1bR96) a fallback from happening if the navigation wasn't originally upgraded by `HttpsOnlyModeUpgradeInterceptor`.

Reply:
> Or maybe just setting strict-transport-security: max-age=31536000; includeSubdomains on https://arthuredelstein.net/? That should turn on the HSTS flag on insecure.arthuredelstein.net too.

I didn't think of that! Yes, that works. I was able to confirm that when `includeSubdomains` is included on https://arthuredelstein.net, when I attempt to visit http://insecure.arthuredelstein.net, I get an interstitial page with `NET::ERR_CERT_COMMON_NAME_INVALID`: 

whereas when `includeSubdomains` is not included and I visit http://insecure.arthuredelstein.net, I see the expected HTTPS upgrade behavior (falling back in Standard mode and showing the HTTPS-Only Mode interstitial in Strict mode).